driver/adfs_proxy.cc

```diff
@@ -28,31 +28,245 @@
+    if (!tagName.empty() && hashSet.find(tagName) == hashSet.end()) {
+      hashSet.insert(tagName);
+      retval.push_back(tag);
+    }
+
+    source = matches.suffix().str();
+  }
+
+  return retval;
+}
+
+std::string ADFS_SAML_UTIL::get_value_by_key(const std::string& input, const std::string& key) {
+  std::string pattern("(");
+  pattern += key;
+  pattern += ")\\s*=\\s*\"(.*?)\"";
+
+  std::smatch matches;
+  if (std::regex_search(input, matches, std::regex(pattern))) {
+    MYLOG_TRACE(init_log_file(), 0, "get_value_by_key");
+    return escape_html_entity(matches.str(2));
+  }
+  return "";
+}
+
+std::string ADFS_SAML_UTIL::get_parameters_from_html(DataSource* ds, const std::string& body) {
+  std::map<std::string, std::string> parameters;
+  for (auto& inputTag : get_input_tags_from_html(body)) {
+    std::string name = get_value_by_key(inputTag, std::string("name"));
+    std::string value = get_value_by_key(inputTag, std::string("value"));
+    std::string nameLower = name;
+    std::transform(nameLower.begin(), nameLower.end(), nameLower.begin(),
+                   [](unsigned char c) { return std::tolower(c); });
+
+    const std::string username = static_cast<const char*>(ds->opt_IDP_USERNAME);
+    const std::string password = static_cast<const char*>(ds->opt_IDP_PASSWORD);
+
+    if (nameLower.find("username") != std::string::npos) {
+      parameters.insert(std::pair<std::string, std::string>(name, username));
+    } else if (nameLower.find("authmethod") != std::string::npos) {
+      if (!value.empty()) {
+        parameters.insert(std::pair<std::string, std::string>(name, value));
+      }
+    } else if (nameLower.find("password") != std::string::npos) {
+      parameters.insert(std::pair<std::string, std::string>(name, password));
+    } else if (!name.empty()) {
+      parameters.insert(std::pair<std::string, std::string>(name, value));
+    }
+  }
+
+  // Convert parameters to a & delimited string, e.g. username=u&password=p
+  const std::string delimiter = "&";
+  const std::string result =
+      std::accumulate(parameters.begin(), parameters.end(), std::string(),
+                      [delimiter](const std::string& s, const std::pair<const std::string, std::string>& p) {
+                        return s + (s.empty() ? std::string() : delimiter) + p.first + "=" + p.second;
+                      });
+
+  return result;
+}
+
+std::string ADFS_SAML_UTIL::get_form_action_body(const std::string& url, const std::string& params) {
+  nlohmann::json res;
+  try {
+    res = this->http_client->post(url, params, "application/x-www-form-urlencoded");
+  } catch (SAML_HTTP_EXCEPTION& e) {
+    const std::string error =
+        "Failed to get SAML Assertion from ADFS : " + e.error_message() + ". Please verify your ADFS credentials.";
+    throw SAML_HTTP_EXCEPTION(error);
+  }
+
+  if (res.empty()) {
+    return "";
+  }
+  return res;
```

Comment:
Very minor and very dependent on style preference. What do you think about using this here?
```C++
return res.empty() ? "" : res;
```